# Part 2: Dimensionality Reduction

**In Part 1, we learnt:**
    
* How to import dataset 
* How to preprocess data
* How to create tokens, vocabulary of known words and TF-IDF matrix
* How to find 10 most significant terms in a class

TFIDF matrix has dimensions of orders of thousands. We know that tfidf matrix is sparse and low-rank.Learning algorithms perform poorly on high-dimensional data. Therefore, we want to reduce the dimensions.

We can transform features into lower-dimensional space via Latent Semantic Indexing and Non-negative Matrix Factorization. We will learn how to apply both of these methods in this part. 

Latent Semantic Indexing(LSI) minimizes the mean squared residual between the original data and the reconstruction from its low-dimensional representation.

You can learn more about LSI [here](https://www.analyticsvidhya.com/blog/2018/10/stepwise-guide-topic-modeling-latent-semantic-analysis/) and [here](https://machinelearningmastery.com/singular-value-decomposition-for-machine-learning/)


**We are going to use Singular Value Decomposition for transforming the matrix into 50 dimensions.**

### Importing Part_1
I first converted Part_1 from ipynb to py. Then I imported it here down below. It works very smoothly. This is so as to avoid writing same code again and again.

In [16]:
from Part_1 import *

In [18]:
vectorizer_tfidf #Cool

<3903x17594 sparse matrix of type '<class 'numpy.float64'>'
	with 190613 stored elements in Compressed Sparse Row format>

## Latent Semantic Indexing using SVD

In [19]:
#Importing SVD
from sklearn.decomposition import TruncatedSVD

We will reduce dimensions of training data of our class: 'Computer Technology'

In [21]:
#Preprocessing training data
training_data=fetch_20newsgroups(subset='train',categories=computer_technology_subclasses,shuffle=True,random_state=42,
                                 remove=('headers','footers','quotes'))
preprocess(training_data.data)

Let's get our TFIDF matrix of training data

In [22]:
counts2=CountVectorizer(min_df=2,stop_words ='english')
X_counts2=counts2.fit_transform(training_data.data)
tfidf_transformer2=TfidfTransformer()
X_tfidf2=tfidf_transformer2.fit_transform(X_counts2)

Let's do for what this part is all about.

In [23]:
#Reducing dimensionality using LSI for min_df=2
svd2=TruncatedSVD(n_components=50,n_iter=10,random_state=42)
svd2.fit(X_tfidf2)
LSI2=svd2.transform(X_tfidf2)
print("The LSI shape when min_df=2 is:")
print(LSI2.shape)

The LSI shape when min_df=2 is:
(2343, 50)


**This means that there are now 2343 documents and 50 most relevant words.**

## Dimensionality reduction using Non-negative Matrix Factorization

**Importing required library**

In [25]:
from sklearn.decomposition import NMF

We have already created TFIDf matrix for training data. We will simply make a model from NMF

In [26]:
model2=NMF(n_components=50,init='random',random_state=0)
W_train2=model2.fit_transform(X_tfidf2)
print("The NMF shape when min_df=2 is:")
print(W_train2.shape)

The NMF shape when min_df=2 is:
(2343, 50)


## Play around and learn

In [28]:
LSI2

array([[ 0.07751021,  0.04337386, -0.08391913, ..., -0.18113037,
         0.01686565,  0.05719849],
       [ 0.1019845 , -0.05958745,  0.00906318, ..., -0.00321237,
         0.02005059,  0.06316133],
       [ 0.10844307,  0.02680134, -0.00158244, ...,  0.02670837,
        -0.02454688,  0.02236188],
       ...,
       [ 0.16934633, -0.010743  , -0.05744238, ...,  0.06469548,
        -0.00476247,  0.02484388],
       [ 0.17604228, -0.04144455, -0.05037223, ...,  0.00850844,
        -0.01332187, -0.09518748],
       [ 0.21284895,  0.02356831,  0.09449639, ..., -0.01797934,
        -0.02382705,  0.07352551]])

In [29]:
pd.DataFrame(LSI2)

0         1         2         3         4         5         6   \
0     0.077510  0.043374 -0.083919 -0.015489 -0.092739 -0.016976  0.064764   
1     0.101985 -0.059587  0.009063  0.025379  0.041841  0.000788 -0.042434   
2     0.108443  0.026801 -0.001582 -0.063800  0.038635 -0.012820 -0.020113   
3     0.121519 -0.059068  0.096189 -0.086386 -0.018306  0.059667  0.068546   
4     0.041004 -0.031618  0.013208 -0.009814  0.019057  0.025633  0.007283   
5     0.260472 -0.076860  0.092562  0.108238  0.003004 -0.066161 -0.120888   
6     0.107004 -0.029320  0.017272  0.006030 -0.022066 -0.029540 -0.087775   
7     0.157656  0.023854 -0.065427 -0.042922 -0.091283 -0.019557 -0.025346   
8     0.131603 -0.071616  0.053008 -0.038905 -0.068898  0.030232  0.041983   
9     0.137235 -0.001010 -0.000593  0.018038 -0.088914 -0.023792 -0.011028   
10    0.290649 -0.134557  0.081882  0.193129 -0.030768 -0.024133 -0.074851   
11    0.257180 -0.051094  0.187498  0.160132 -0.033920  0.020712  0.101807   
12    0.128139  0.150350  0.079696 -0.021726  0.023293 -0.062119  0.043512   
13    0.138252 -0.018737  0.033084 -0.001929 -0.045661 -0.040358  0.000633   
14    0.142030 -0.058504  0.031831  0.001128 -0.016989 -0.053940 -0.072385   
15    0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
16    0.202887 -0.064263 -0.018925 -0.081703  0.010214 -0.007149 -0.108214   
17    0.090752  0.004836 -0.020423 -0.064008 -0.019011 -0.031001 -0.022630   
18    0.216519 -0.023442 -0.240745  0.063276  0.170814  0.039346  0.085784   
19    0.183016  0.100880  0.107107 -0.065430  0.055861 -0.016508  0.040722   
20    0.169836 -0.041247  0.017167  0.013080 -0.067471 -0.059508 -0.044306   
21    0.051015 -0.004002 -0.012525  0.015443 -0.026702  0.012009  0.032202   
22    0.124952 -0.022794 -0.029481 -0.036394 -0.047027 -0.064732 -0.032386   
23    0.144617 -0.043927  0.057473  0.068896 -0.101738 -0.074611  0.030972   
24    0.185273 -0.131909  0.065244 -0.191555  0.128853  0.083325 -0.012472   
25    0.112124  0.022382 -0.125829  0.028101 -0.025081  0.058718  0.068434   
26    0.088086  0.005821 -0.015642 -0.023940 -0.069983 -0.013340  0.003204   
27    0.104309  0.070863 -0.072967  0.056780 -0.080462  0.110780 -0.004409   
28    0.154965 -0.066690  0.092687  0.011945 -0.094963  0.039359  0.048517   
29    0.133514 -0.062719  0.008684 -0.134890  0.013277  0.055661 -0.033862   
...        ...       ...       ...       ...       ...       ...       ...   
2313  0.117207 -0.058232  0.041308  0.128858  0.007050 -0.045571 -0.126744   
2314  0.226876  0.050135  0.085143  0.091069  0.004481  0.028831  0.037354   
2315  0.053864  0.032153 -0.000419  0.005883 -0.047127 -0.056081  0.002578   
2316  0.104295 -0.057775  0.009217 -0.075598 -0.033003  0.051430  0.039719   
2317  0.156659 -0.040971  0.005956  0.010094 -0.076612 -0.023620 -0.102677   
2318  0.230999 -0.063754 -0.008398  0.069671  0.126389  0.005356 -0.047870   
2319  0.135422 -0.047416 -0.201175  0.122412  0.240342  0.009315  0.049947   
2320  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2321  0.178482 -0.032780  0.033205  0.029742 -0.078287 -0.021281  0.043688   
2322  0.106211  0.006169 -0.014172 -0.012083 -0.013180 -0.036967 -0.003322   
2323  0.096418  0.008804 -0.023392 -0.003339 -0.098420  0.001090 -0.002834   
2324  0.179511  0.028094  0.044071  0.103965 -0.060032 -0.064235 -0.047099   
2325  0.167310 -0.014446 -0.026722 -0.068356 -0.051202 -0.044279 -0.099627   
2326  0.227489  0.021475 -0.134329  0.073200 -0.029101  0.030842  0.098408   
2327  0.152970 -0.003871 -0.010599  0.012756 -0.084617 -0.073148 -0.038388   
2328  0.152132 -0.079669  0.027485  0.060514 -0.025208 -0.027818 -0.079180   
2329  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2330  0.124344  0.009049 -0.037050 -0.002477 -0.079216 -0.037894 -0.086504   
2331  0.172535 -0.006535 -0.081973  0.054689  0.038460 -0.087200 -0.146167   
2332  0.165446 -0.031430  0.003964